src/backend/renderer/gles/profiler.rs

```diff
@@ -0,0 +1,186 @@
+
+        unsafe {
+            gl.QueryCounterEXT(tracepoint.entered, ffi::TIMESTAMP_EXT);
+        }
+
+        let span = self.context.span(span_location).unwrap();
+
+        EnteredGpuTracepoint {
+            span: Some(GpuSpan { tracepoint, span }),
+        }
+    }
+
+    pub fn exit(&mut self, gl: &ffi::Gles2, mut exited: EnteredGpuTracepoint) {
+        let Some(mut span) = exited.span.take() else {
+            return;
+        };
+
+        unsafe {
+            gl.QueryCounterEXT(span.tracepoint.exited, ffi::TIMESTAMP_EXT);
+        }
+        span.span.end_zone();
+
+        self.tracepoints.push(ExitedGpuTracepoint { span });
+    }
+
+    pub fn collect(&mut self, gl: &ffi::Gles2) {
+        let mut i = 0;
+        while i != self.tracepoints.len() {
+            let mut available: ffi::types::GLuint = 0;
+            unsafe {
+                gl.GetQueryObjectuivEXT(
+                    self.tracepoints[i].span.tracepoint.exited,
+                    ffi::QUERY_RESULT_AVAILABLE,
+                    &mut available,
+                );
+                if gl.GetError() != ffi::NO_ERROR {
+                    self.tracepoints.remove(i);
+                    continue;
+                }
+            }
+
+            if available == 1 {
+                let tracepoint = self.tracepoints.remove(i);
+                let mut start_timestamp = 0;
+                let mut end_timestamp = 0;
+                unsafe {
+                    gl.GetQueryObjecti64vEXT(
+                        tracepoint.span.tracepoint.entered,
+                        ffi::QUERY_RESULT,
+                        &mut start_timestamp,
+                    );
+                    gl.GetQueryObjecti64vEXT(
+                        tracepoint.span.tracepoint.exited,
+                        ffi::QUERY_RESULT,
+                        &mut end_timestamp,
+                    );
+                }
+                tracepoint
+                    .span
+                    .span
+                    .upload_timestamp(start_timestamp, end_timestamp);
+            } else {
+                i += 1;
+            }
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+struct GpuTracepoint {
+    entered: ffi::types::GLuint,
+    exited: ffi::types::GLuint,
+}
+
+struct GpuSpan {
+    tracepoint: GpuTracepoint,
+    span: tracy_client::GpuSpan,
+}
+
+pub struct EnteredGpuTracepoint {
+    span: Option<GpuSpan>,
+}
+
+pub struct ScopedGpuTracepoint<'a, 'b> {
+    span: Option<EnteredGpuTracepoint>,
+    profiler: &'a mut GpuProfiler,
+    gl: &'b ffi::Gles2,
+}
+
+impl<'a, 'b> Drop for ScopedGpuTracepoint<'a, 'b> {
+    fn drop(&mut self) {
+        if let Some(span) = self.span.take() {
+            self.profiler.exit(self.gl, span);
+        }
+    }
+}
+
+struct ExitedGpuTracepoint {
+    span: GpuSpan,
+}
+
+impl Drop for EnteredGpuTracepoint {
+    fn drop(&mut self) {
+        debug_assert!(self.span.is_none());
```

Comment:
Doing things this way clashes with the use of `?` for error handling; it will panic if `?` prevents it from getting to the `exit` call.

Not sure what to do about this.